### PR TITLE
Fix the crash at the end (issue #476)

### PR DESCRIPTION
**Pull request by einola**
# Fix a crash at the end
### Task List
- [ ] Defined the tests that specify a complete and functioning change (*It may help to create a [design specification & test specification](../../../wiki/Specification-Template)*)
- [ ] Implemented the source code change that satisfies the tests
- [ ] Documented the feature by providing worked example
- [ ] Updated the README or other documentation
- [ ] Completed the pre-Request checklist below

---
# Change Description

I fix the crash at the end by replacing the destructor of Timer (in Timer.hpp). This seems to work, but I wonder if there are side effects because there are two constructors, Timer() and Timer(const Key& rootKey), and because the destructor is no longer a virtual function.

---
# Test Description

Run cmake with CMAKE_BUILD_TYPE=Release, and compile the model from scratch. Run either the thermodynamics or june23 integration tests. These should run to the end, but end with a segmentation fault (signal 11).

---
# Documentation Impact

None

---
# Other Details

None

---
### Pre-Request Checklist

- [ ] The requirements of this pull request are fully captured in an issue or design specification and are linked and summarised in the description of this PR
- [ ] No new warnings are generated
- [ ] The documentation has been updated (or an issue has been created to track the corresponding change)
- [ ] Methods and Tests are commented such that they can be understood without having to obtain additional context
- [ ] This PR/Issue is labelled as a bug/feature/enhancement/breaking change
- [ ] File dates have been updated to reflect modification date
- [ ] This change conforms to the conventions described in the README

### CODE REVIEW

**Comment by timspainNERSC** [comment]
Is it the lack of `virtual` or the `reset()` call that fixes the crash? Or are both changes necessary?

**Comment by einola** [comment]
Good point! I thought it was the `reset()` call that fixed things, but it's in fact removing `virtual` that does it. I.e. 

```
    ~Timer() = default;
```

works, and so does

```
    ~Timer() { reset(); }
```

but both

```
    virtual ~Timer() { reset(); }
```

and

```
    virtual ~Timer() { reset(); }
```

give a crash at the end. Which also means that I don't understand what's going on. I thought that calling `reset()` would clear the memory so that we could have a clean exit.

**Comment by timspainNERSC** [comment]
> Good point! I thought it was the `reset()` call that fixed things, but it's in fact removing `virtual` that does it.

That does suggest a deeper problem with how I am using the class. I'll take a look and see if I can see anything obviously wrong. If I can't find anything, then we can just push the non-`virtual` version.

File: core/src/include/Timer.hpp [hunk 1]
```diff
@@ -39,7 +39,7 @@ class Timer {
      * @param rootKey Name of the root node.
      */
     Timer(const Key& rootKey);
-    virtual ~Timer() = default;
+    ~Timer() { reset(); }
```
Comment (einola) [inline core/src/include/Timer.hpp:42]:
```suggestion
    ~Timer() = default;
```

**Comment by timspainNERSC** [comment]
> > Good point! I thought it was the `reset()` call that fixed things, but it's in fact removing `virtual` that does it.
> 
> That does suggest a deeper problem with how I am using the class. I'll take a look and see if I can see anything obviously wrong. If I can't find anything, then we can just push the non-`virtual` version.

Not a flippin´scooby.

**Review by timspainNERSC — APPROVED** [review]
Let's get this merged.

**Pull request merged**